### PR TITLE
Support 2-d broadcast for observe

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

We infer the type of broadcasting in `args` vs. array of `args` and `spin_op` vs. array of `spin_op` when handling Python `observe`.
However, we didn't handle a combination of them. 

This PR:

- Allow Python `observe` to return a nested vector. Peel off the `spin_op` sum result to construct the nested result.

- Add a unit test case.

Resolved https://github.com/NVIDIA/cuda-quantum/issues/776

### CODE REVIEW

**Review by amccaskey — APPROVED** [review]
Thanks for the fix

File: python/runtime/cudaq/algorithms/py_observe.cpp [hunk 1]
```diff
@@ -302,6 +303,20 @@ void bindObserve(py::module &mod) {
         }
 
         // Return the vector of results, this is for observe_n
```
Comment (bmhowe23) [inline python/runtime/cudaq/algorithms/py_observe.cpp:305]: nit: I think this pre-existing `// Return the vector of results, this is for observe_n` comment should be moved below your new code to be kept with the `return {result}` statement.

**Review by bmhowe23 — APPROVED** [review]
LGTM

**Review by schweitzpgi — APPROVED** [review]
👍

**Pull request merged**